ladybug_geometry/geometry3d/face.py

```diff
@@ -0,0 +1,1239 @@
+# coding=utf-8
+"""Planar Face in 3D Space"""
+from __future__ import division
+
+from .pointvector import Point3D, Point3DImmutable, Vector3D, Vector3DImmutable
+from .line import LineSegment3D, LineSegment3DImmutable
+from .plane import Plane
+from .mesh import Mesh3D
+from ._2d import Base2DIn3D
+
+from ..intersection3d import closest_point3d_on_line3d
+
+from ..geometry2d.pointvector import Vector2D
+from ..geometry2d.ray import Ray2D
+from ..geometry2d.polygon import Polygon2D
+from ..geometry2d.mesh import Mesh2D
+
+import math
+try:
+    from itertools import izip as zip  # python 2
+except ImportError:
+    xrange = range  # python 3
+
+
+class Face3D(Base2DIn3D):
+    """Planar Face in 3D space.
+
+    Properties:
+        vertices
+        plane
+        polygon2d
+        triangulated_mesh2d
+        triangulated_mesh3d
+        boundary
+        holes
+        boundary_segments
+        hole_segments
+        normal
+        min
+        max
+        center
+        perimeter
+        area
+        centroid
+        is_clockwise
+        is_convex
+        is_self_intersecting
+        has_holes
+    """
+    __slots__ = ('_vertices', '_plane', '_polygon2d', '_triangulated_mesh2d',
+                 '_triangulated_mesh3d', '_boundary', '_holes',
+                 '_boundary_segments', '_hole_segments'
+                 '_min', '_max', '_center', '_perimeter', '_area', '_centroid',
+                 '_is_clockwise', '_is_convex', '_is_complex')
+
+    def __init__(self, vertices, plane):
+        """Initilize Face3D.
+
+        Args:
+            vertices: A list of Point3D objects representing the vertices of the sface.
+            plane: A Plane object indicating the plane in which the face exists.
+        """
+        self._check_vertices_input(vertices)
+        assert isinstance(plane, Plane), 'Expected Plane for Face3D.' \
+            ' Got {}.'.format(type(plane))
+        self._plane = plane
+
+        self._polygon2d = None
+        self._mesh2d = None
+        self._mesh3d = None
+        self._boundary = self._vertices
+        self._holes = None
+        self._boundary_segments = None
+        self._hole_segments = None
+
+        self._min = None
+        self._max = None
+        self._center = None
+        self._perimeter = None
+        self._area = None
+        self._centroid = None
+        self._is_clockwise = None
+        self._is_convex = None
+        self._is_complex = None
```

Comment:
do we have `is_complex` documented somewhere as what does it mean?

Reply:
`_is_complex` is exposed as the `is_self_intersecting` property, meaning that the shape intersects with itself like a bow tie shape.  complex vs. simple are the mathematical terms to describe this but I exposed it as is_self_intersecting because it is a lot clearer.  I'll change the name of the hidden property to match the one that is exposed.  This will make it clearer to people editing the code.